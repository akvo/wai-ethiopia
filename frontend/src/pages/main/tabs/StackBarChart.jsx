import React, { useState, useMemo } from 'react';
import { Spin } from 'antd';
import { LoadingOutlined } from '@ant-design/icons';
import uniq from 'lodash/uniq';
import sum from 'lodash/sum';
import Chart from '../../../chart';
import PaginationApi from '../../../components/PaginationApi';

const PER_PAGE = 250;

const getAdmScores = (chartScores, data) => {
  // Initial value of each option within each administration
  const statusValues = {};
  const statusPercentages = {};
  const scoreValues = {};
  uniq(data.map((d) => d?.administration))?.forEach((d) => {
    const dataItems = data?.[0]?.child?.map((c) => ({
      [c?.option]: 0,
    }));
    const initialValues = Object.assign({}, ...dataItems);
    statusValues[d] = initialValues;
    statusPercentages[d] = initialValues;
    scoreValues[d] = [];
  });
  for (let i = 0; i < data.length; i++) {
    const { administration, child } = data[i];
    for (let j = 0; j < child.length; j++) {
      const { option, percent } = child[j];
      statusValues[administration][option] += percent;
    }
  }
  // Calculate the percentage of each option within each administration
  for (const administration in statusValues) {
    const totalPercent = Object.values(statusValues[administration]).reduce(
      (total, percent) => total + percent,
      0
    );
    for (const option in statusValues[administration]) {
      const percent =
        (statusValues[administration][option] / totalPercent) * 100;
      const score = chartScores[option] || 0;
      scoreValues[administration].push((score * percent) / 100);
      statusPercentages[administration][option] = percent || 0;
    }
  }
  return {
    statusPercentages,
    scoreValues,
  };
};

const StackBarChart = ({
  apiUrl,
  chartScores,
  totalPages,
  data,
  selectedAdministration,
}) => {
  const initLoading = totalPages > 1 ? true : false;
  const [chartValues, setChartValues] = useState(data);
  const [loading, setLoading] = useState(initLoading);
  const [showEmptyValueChart, setShowEmptyValueChart] = useState(false);

  const paginationCallback = (allData) => {
    const { statusPercentages, scoreValues } = getAdmScores(
      chartScores,
      allData
    );
    const _chartValues = data.map((d) => ({
      ...d,
      score: sum(scoreValues?.[d.id]),
      stack: d?.stack?.map((st) => ({
        ...st,
        value: statusPercentages?.[d.id]?.[st?.name],
      })),
    }));
    setChartValues(_chartValues);
    setLoading(false);
  };

  const filteredChartValues = useMemo(() => {
    if (showEmptyValueChart) {
      return chartValues;
    }
    const filterTmp = chartValues
      .map((d) => {
        const filterStack = d.stack.filter((s) => s.value > 0 || s.value);
        if (!filterStack.length) {
          return false;
        }
        return {
          ...d,
          stack: filterStack,
        };
      })
      .filter((x) => x);
    return filterTmp;
  }, [showEmptyValueChart, chartValues]);

  const height = filteredChartValues.length * 60;

  return (
    <div className="jmp-chart-container">
      <PaginationApi
        apiUrl={apiUrl}
        totalPages={totalPages}
        perPage={PER_PAGE}
        callback={paginationCallback}
      />
      <div className="jmp-chart">
        {loading ? (
          <div className="chart-loading">
            <Spin indicator={<LoadingOutlined />} />
          </div>
        ) : (
          <Chart
            title=""
            subTitle=""
            type={'JMP-BARSTACK'}
            data={filteredChartValues}
            wrapper={false}
            height={height}
            extra={{
              selectedAdministration,
            }}
            emptyValueCheckboxSetting={{
              show: true,
              checked: showEmptyValueChart,
              handleOnCheck: () => {
                setLoading(true);
                setShowEmptyValueChart(!showEmptyValueChart);
                setTimeout(() => {
                  setLoading(false);
                }, 500);
              },
            }}
          />
        )}
      </div>
    </div>
  );
};

export default StackBarChart;
