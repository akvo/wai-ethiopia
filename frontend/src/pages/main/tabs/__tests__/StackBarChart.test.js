import React from 'react';
import renderer from 'react-test-renderer';
import { render, waitFor, act } from '@testing-library/react';
import { unmountComponentAtNode } from 'react-dom';
import 'jest-canvas-mock';
import StackBarChart from '../StackBarChart';

const props = {
  apiUrl: 'chart/jmp-data/1/water?administration=0',
  chartScores: { Basic: 10, Limited: 1, 'No service': -1 },
  data: [
    {
      id: 1,
      name: 'Barguna',
      parent: null,
      score: null,
      stack: [
        {
          color: '#ffda46',
          id: 0,
          name: 'No service',
          order: 1,
          value: 50,
        },
        {
          color: '#fff176',
          id: 1,
          name: 'Limited',
          order: 2,
          value: 15,
        },
        { color: '#00b8ec', id: 2, name: 'Basic', order: 3, value: 35 },
      ],
    },
    {
      id: 2,
      name: 'Satkhira',
      parent: null,
      score: null,
      stack: [
        {
          color: '#ffda46',
          id: 0,
          name: 'No service',
          order: 1,
          value: 50,
        },
        {
          color: '#fff176',
          id: 1,
          name: 'Limited',
          order: 2,
          value: 0,
        },
        { color: '#00b8ec', id: 2, name: 'Basic', order: 3, value: 50 },
      ],
    },
  ],
  height: 100,
  name: 'Water',
  question: 'water',
  selectedAdministration: null,
  totalPages: 1,
};

const paginationData = {
  current: 2,
  data: [
    {
      administration: 1,
      child: [
        {
          option: 'No service',
          count: 0,
          percent: 0,
          color: '#ffda46',
        },
        {
          option: 'Limited',
          count: 3,
          percent: 20.0,
          color: '#fff176',
        },
        {
          option: 'Basic',
          count: 11,
          percent: 73.33333333333333,
          color: '#00b8ec',
        },
      ],
    },
    {
      administration: 2,
      child: [
        {
          option: 'No service',
          count: 0,
          percent: 0,
          color: '#ffda46',
        },
        {
          option: 'Limited',
          count: 37,
          percent: 23.417721518987342,
          color: '#fff176',
        },
        {
          option: 'Basic',
          count: 99,
          percent: 62.65822784810127,
          color: '#00b8ec',
        },
      ],
    },
  ],
  total: 423,
  total_page: 2,
  question: 'water',
  scores: [
    ['No service', -1],
    ['Limited', 1],
    ['Basic', 10],
  ],
};

jest.mock('../../../../util/api', () => {
  return {
    get: () => new Promise((resolve) => resolve({ data: paginationData })),
  };
});

const mockChartComponent = jest.fn();
// eslint-disable-next-line react/display-name
jest.mock('../../../../chart', () => (props) => {
  mockChartComponent(props);
  return <mock-chartComponent />;
});

describe('StackBarChart', () => {
  let container = null;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
  });

  it('should render correctly & match snapshot', () => {
    const component = renderer.create(<StackBarChart {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not call pagination callback when totalPages = 1', () => {
    const paginationCallback = jest.fn();
    render(<StackBarChart {...props} />, container);
    expect(paginationCallback).toHaveBeenCalledTimes(0);
  });

  it('should call pagination callback when totalPages > 1', async () => {
    const paginationCallback = jest.fn();
    await act(async () =>
      render(<StackBarChart {...{ ...props, totalPages: 2 }} />, container)
    );
    act(() => {
      paginationCallback();
    });
    await waitFor(() => {
      expect(paginationCallback).toHaveBeenCalledTimes(1);
      expect(mockChartComponent).toHaveBeenCalledWith(
        expect.objectContaining({
          data: [
            {
              id: 1,
              name: 'Barguna',
              parent: null,
              score: 8.071428571428571,
              stack: [
                {
                  color: '#ffda46',
                  id: 0,
                  name: 'No service',
                  order: 1,
                  value: 0,
                },
                {
                  color: '#fff176',
                  id: 1,
                  name: 'Limited',
                  order: 2,
                  value: 21.42857142857143,
                },
                {
                  color: '#00b8ec',
                  id: 2,
                  name: 'Basic',
                  order: 3,
                  value: 78.57142857142857,
                },
              ],
            },
            {
              id: 2,
              name: 'Satkhira',
              parent: null,
              score: 7.551470588235295,
              stack: [
                {
                  color: '#ffda46',
                  id: 0,
                  name: 'No service',
                  order: 1,
                  value: 0,
                },
                {
                  color: '#fff176',
                  id: 1,
                  name: 'Limited',
                  order: 2,
                  value: 27.205882352941174,
                },
                {
                  color: '#00b8ec',
                  id: 2,
                  name: 'Basic',
                  order: 3,
                  value: 72.79411764705883,
                },
              ],
            },
          ],
          extra: { selectedAdministration: null },
          height: 320,
          subTitle: '',
          title: '',
          type: 'JMP-BARSTACK',
          wrapper: false,
        })
      );
    });
  });
});
