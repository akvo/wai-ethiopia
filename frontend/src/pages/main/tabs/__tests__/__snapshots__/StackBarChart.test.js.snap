// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackBarChart should render correctly & match snapshot 1`] = `
<div
  className="jmp-chart-container"
>
  <div
    className="jmp-chart"
  >
    <mock-chartComponent />
  </div>
</div>
`;
